### ISSUE TITLE
undefined method `name' for nil:NilClass in "app/views/public_activity/repository/_push.html.slim"

### ISSUE THREAD

**Issue by sotsy** [SUSE/Portus#338]
If a new push event appears in Portus/Admin I get an `undefined method `name' for nil:NilClass` error.

I think it happens if `activity.recipient.name` is nil and in db is no recipient_id

`app/views/public_activity/repository/_push.html.slim` related line:

```
= link_to "#{activity.trackable.name}:#{activity.recipient.name}", activity.trackable
```

A tracked activity from `activities` table looks like the following

```
| id | trackable_id | trackable_type | owner_id | owner_type | key             | parameters        | recipient_id | recipient_type | created_at          | updated_at          |
|  3 |            1 | Repository     |        4 | User       | repository.push | NULL              |         NULL | Tag            | 2015-09-15 14:17:19 | 2015-09-15 14:17:19 |
```

**Comment by sotsy** [comment]
In addition to that `Showing /portus/app/views/repositories/show.html.slim where line #35 raised:`

```
undefined method `username' for nil:NilClass
```

```
td= tag.author.username
```

**Comment by flavio** [comment]
How can this bug be reproduced? We don't understand how the DB could have ended up in this state.

**Comment by morodin** [comment]
We set-up our environment using LDAP and because of this, TLS for both the registry and Portus. Next, we logged in two users: the first was made an admin, the second an ordinary user.
After the default namespace was created for this user, we used docker client on a remote machine to log into the registry, tag an image and push it to the registry:

```
docker login -u userx -e "userx@our.domain.com" registry.our.domain.com:5000
docker tag docker.io/busybox registry.our.domain.com:5000/userx/busybox
docker push registry.our.domain.com:5000/userx/busybox
``` 

This created the errors above - first the _push.html.slim error - immediately after the notification from the registry was received by Portus and either user "userx" or an admin revisited the Portus dashboard. 

The second error occured after the crono-job updating Portus using the catalog was run and the UI was visited again.

**Comment by mssola** [comment]
I cannot reproduce this issue. I've followed the exact same steps as @morodin and it works fine for me. So, some questions:

1. Which version of Portus are you using? Is it from the appliance, master, ... ?
2. Have you touched the DB manually or in any way in which the DB is now in an inconsistent state?
3. Could you provide logs ?

Thanks.

**Comment by morodin** [comment]
We used the Portus version from the master, just doing as described at the bottom from "http://suse.github.io/Portus/" (besides doing a compose-setup.sh). The code we got was from September 8th. 

Before submitting the bug report, we removed all databases and redid the compose-setup. The database was empty and we did not manually change it - just did a few "select * from .. ". 

Of course we can provide log. We can also add any additional debugging. What do you need?

**Comment by mssola** [comment]
@morodin just the `log/development.log` file should be enough if you're using the compose setup. Moreover, just in case, could you also provide the `log/crono.log` file ?

Thanks!

**Comment by morodin** [comment]
I restarted out containers, logged into the UI, logged in with the dogger client and pushed another image. I uploaded the following log files to 
https://owncloud.fokus.fraunhofer.de/index.php/s/DGZigLkyfExFD4r

150917_registry.log - the log file from the registry.
150917_development.log - the development log file.
150917_db_output.log - the relevant part of the db
150917_crono.output - the output from the crono task.

Do you need anything else?

**Comment by morodin** [comment]
Oh, i forgot - we changed the slim files so that the error does not occur in the moment because we have the server already under inspection by our users. Do you need  us to reset the changes and run the log again with the original code?

**Comment by mssola** [comment]
@morodin sorry to bother you, but yes, I'd appreciate if you could run the original code and pass me the logs. For what you've showed me, the DB looks inconsistent, that's I was asking previously if you had touched the DB manually. Thanks for your patience.

**Comment by plamer** [comment]
Hey there, I've hit the same issue (I think) - I'm running the appliance, but I've replaced the docker repository with version 2.1.1. Here is the production.log file:

    http://pastebin.com/Q60UKtic

One admin user, created second user, logged in, tag, push and then this happened. No code is altered from the application.

**Comment by flavio** [comment]
I guess you are not using LDAP. Am I right?

**Comment by plamer** [comment]
Oh, no, I'm not - sorry for missing that

**Comment by morodin** [comment]
We commented our changes to the slim files and did the test again. You will find the logs in
https://owncloud.fokus.fraunhofer.de/index.php/s/uIdSDwnqIR7LKCK

In the development log are 4 places marked with "XXX":
* "First Error - Correcting" - after the first error, before the correction of the first slim file,
* "Calling crono job" - before calling the crono jub,
* "correcting second error" - after the second error, before correcting it
* "done" - at the end of the test.

We can see the following behaviour: After the image is pushed to the registry, the "Tag page" is empty, it gets filled with the "latest" flag only after the crono job was run.

**Comment by mssola** [comment]
@morodin sorry for the delay. Your last comment made me suspicious, I re-read the discussion again and I realized that you are using "master" from 8th September. Is that correct? If that's the case, note that there was a known regression that got fixed [here](https://github.com/SUSE/Portus/pull/318). That's why you're experiencing the tags only appearing after the crono job  kicks in. This leads to an unfortunate situation in which the DB gets inconsistent.

Could you update to the current master and try to reproduce this again ?

**Comment by morodin** [comment]
@mssola: I pulled the latest from master and ran into a lot of problems - so i cant really comment on wether the issue is fixed.
We are using TLS both on the registry as well the GUI. In the current master files, you use the RegistryClient, but alway with "use_ssl = false". In this case we get the following error message
```Could not fetch the tag for target {"mediaType"=>"application/vnd.docker.distribution.manifest.v1+json", "size"=>6709, "digest"=>"sha256:621382a9bb17b9dd62d44cb054db885de3001b99ea1eafba709915251005ad90", "length"=>6709, "repository"=>"userx/busybox8", "url"=>"https://dregistry.example.com:5000/v2/userx/busybox8/manifests/sha256:621382a9bb17b9dd62d44cb054db885de3001b99ea1eafba709915251005ad90"}.```

If we change this, so that the client uses ssl (btw - shouldn't this be a configuration option?), the process gets stuck (the output is from debug messages we included in the http helper):
```
get_response_token: uri=dregistry.example.com,5000, use_ssl=true
calling http.request get_response_token: uri=dregistry.example.com,5000, use_ssl=true
get_response_token: uri=dregistry.example.com,3000, use_ssl=true
```

**Comment by mssola** [comment]
@morodin the fact that the registry client doesn't care about SSL is a known limitation, and there should be an issue about that (I'm opening up one right now). I'll try to fix it as soon as possible and then you could try to update again and see if these issues are gone for you.

**Comment by morodin** [comment]
Great, thanks :) Hardwiring it worked as a work-around in the last release we used - but to have it handled correctly would be fabulous.

**Comment by mssola** [comment]
@morodin after the PR #363, it should work now. Could you update to current master and tell us if you can still reproduce this issue ? Thanks for your patience :)

**Comment by morodin** [comment]
After a quick check it still hangs in the get_reponse_token call. I will try again tomorrow with a clean db and give you feedback.

**Comment by aleksnikolic** [comment]
Hi guys,

I have encountered the same error as described here. I didn't use compose setup script as I am using the database on the remote machine. Portus version 1.0.1 got from the release tarball here on github. Portus and registry are both behind the reverse proxy (nginx and varnish).

Everything was set up correctly, I was able to login into UI and on a remote docker daemon but as soon as I pushed the first image to the registry it drops this error and I am not able to access the UI. 

```ruby
Repository Exists (0.7ms)  SELECT  1 AS one FROM `repositories` WHERE (`repositories`.`name` = BINARY 'bamboo-java-client' AND `repositories`.`id` != 1 AND `repositories`.`namespace_id` = 1) LIMIT 1
  SQL (0.7ms)  UPDATE `repositories` SET `namespace_id` = 1, `updated_at` = '2015-09-24 16:37:18' WHERE `repositories`.`id` = 1
   (27.1ms)  COMMIT
Completed 202 Accepted in 316ms (ActiveRecord: 67.5ms)


Started GET "/admin/dashboard" for 172.17.0.2 at 2015-09-24 16:37:25 +0000
Cannot render console from 172.17.0.2! Allowed networks: 127.0.0.1, ::1, 127.0.0.0/127.255.255.255
Processing by Admin::DashboardController#index as HTML
  User Load (0.9ms)  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 1  ORDER BY `users`.`id` ASC LIMIT 1
   (0.8ms)  SELECT COUNT(*) FROM `registries`
   (0.6ms)  SELECT COUNT(*) FROM `users`
   (0.6ms)  SELECT COUNT(*) FROM `teams`
   (0.6ms)  SELECT COUNT(*) FROM `namespaces`
   (0.7ms)  SELECT COUNT(count_column) FROM (SELECT  1 AS count_column FROM `activities` LIMIT 20) subquery_for_count
  PublicActivity::Activity Load (0.7ms)  SELECT  `activities`.* FROM `activities`  ORDER BY created_at DESC LIMIT 20
  User Load (0.7ms)  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 1 LIMIT 1
  Repository Load (0.7ms)  SELECT  `repositories`.* FROM `repositories` WHERE `repositories`.`id` = 1 LIMIT 1
  Namespace Load (0.6ms)  SELECT  `namespaces`.* FROM `namespaces` WHERE `namespaces`.`id` = 1 LIMIT 1
  Registry Load (0.6ms)  SELECT  `registries`.* FROM `registries` WHERE `registries`.`id` = 1 LIMIT 1
  Rendered public_activity/repository/_push.html.slim (21.2ms)
  Rendered admin/dashboard/index.html.slim within layouts/application (115.8ms)
Completed 500 Internal Server Error in 131ms (ActiveRecord: 8.9ms)

ActionView::Template::Error (undefined method `name' for nil:NilClass):
    10:           = "#{activity.owner.username} pushed "
    11:           = link_to activity.trackable.namespace.clean_name, activity.trackable.namespace
    12:           | /
    13:           = link_to "#{activity.trackable.name}:#{activity.recipient.name}", activity.trackable
  app/views/public_activity/repository/_push.html.slim:13:in `_app_views_public_activity_repository__push_html_slim__1131702872235109430_70003304494820'
  app/views/admin/dashboard/index.html.slim:58:in `block in _app_views_admin_dashboard_index_html_slim___557323738472753169_70003267681340'
  app/views/admin/dashboard/index.html.slim:57:in `_app_views_admin_dashboard_index_html_slim___557323738472753169_70003267681340'


  Rendered /usr/local/bundle/gems/actionpack-4.2.2/lib/action_dispatch/middleware/templates/rescues/_source.erb (10.9ms)
  Rendered /usr/local/bundle/gems/actionpack-4.2.2/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb (4.1ms)
  Rendered /usr/local/bundle/gems/actionpack-4.2.2/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb (1.1ms)
  Rendered /usr/local/bundle/gems/actionpack-4.2.2/lib/action_dispatch/middleware/templates/rescues/template_error.html.erb within rescues/layout (53.8ms)
```
Can you tell me which Portus version (git tag or commit-sha) this issue is addressed? If you need more details please let me know.

**Comment by mssola** [comment]
Hi @aleksnikolic, this could've been fixed by either 09dc052721ad560e4fb56c585f6193d3305fc578 or 702f42644c31e444267faf228b5bc91de6f26ef4. Note that Portus 1.0.1 was realeased quite some time ago, so I'd recommend you to just update to master :)

**Comment by mssola** [comment]
Hi @morodin, can you still reproduce this issue with a clean db ?

**Comment by morodin** [comment]
Hi,
i can confirm that this first issue is gone (the one in _push.html.slim). However, this second issue is still "active" (the one in '/portus/app/views/repositories/show.html.slim')

I can also see problems with the 'tag' handling. When adding a new tag to a image, it will get named as '1', which is later corrected when running the crono job:

Also, i still have the issue that the rails app gets frozen. When i push an image, i get the following output (including debug output from us):

```
Started POST "/v2/webhooks/events" for 172.17.42.1 at 2015-09-25 10:59:50 +0000
 Cannot render console from 172.17.42.1! Allowed networks: 127.0.0.1, ::1, 127.0.0.0/127.255.255.255
 Processing by Api::V2::EventsController#create as JSON
   Registry Load (0.4ms)  SELECT  `registries`.* FROM `registries` WHERE `registries`.`hostname` = 'dregistry.example.com:5000' LIMIT 1
   Namespace Load (0.3ms)  SELECT  `namespaces`.* FROM `namespaces` WHERE `namespaces`.`registry_id` = 1 AND `namespaces`.`name` = 'userx' LIMIT 1
 Registry client called, host=dregistry.example.com:5000, use_ssl=true, username=
 Registry client manifest, repoistory=userx/busybox5, tag=sha256:eb026ddcca640a69c66785995807cf2ba5c53347a10ab4debfdedbc8b138e264
 get_response_token: uri=dregistry.example.com,5000, use_ssl=true
 calling http.request get_response_token: uri=dregistry.example.com,5000, use_ssl=true
 done http request
 get_response_token: uri=dregistry.example.com,3000, use_ssl=true

```
Could this be related to a problem because of running the crono-job inside the container? The frozen application makes it really hard to debug.

**Comment by aleksnikolic** [comment]
@mssola Many thanks, That's working but I have another problem now. I am currently running the latest portus version from master and according to the logs portus cannot retrieve the tag from registry. I know that registry have changed the format of the event so the tag is part of the URL but I also thought you guys fixed it from the portus side.

Sorry as this is not related to the original issue in this thread. I can open a new one if needed.

```ruby
Could not fetch the tag for target {"mediaType"=>"application/vnd.docker.distribution.manifest.v1+json", "size"=>54135, "digest"=>"sha256:b8f0a377495e6ec953d25547363a7a5d8560bb2e75f92e4dd44cc7ac28af645d", "length"=>54135, "repository"=>"aleks/httpd", "url"=>"https://distribution.test.lan/v2/aleks/httpd/manifests/sha256:b8f0a377495e6ec953d25547363a7a5d8560bb2e75f92e4dd44cc7ac28af645d"}
```

**Comment by mssola** [comment]
@aleksnikolic have you set Portus to use SSL for the communication between the application and the registry ? Notice that you now have a way to set it in the admin panel as specified in the PR #363. Otherwise, could you provide logs ? Thanks.

**Comment by aleksnikolic** [comment]
@mssola Thanks for reply. Toggling ssl on and off didn't make any difference. The setup is like this: Both registry and portus from user perspective is behind the reverse proxy. The event notifications are sent via direct docker link between registry and Portus, but Portus connection to the registry is going via reverse proxy that terminates the SSL connection. I did tcpdump and according to the registry logs every time the push is successful and Portus got notified and when it tries to get the manifest from the registry it encounters the Authorisation error 

```ruby
{"errors":[{"code":"UNAUTHORIZED","message":"access to the requested resource is not authorized","detail":[{"Type":"repository","Name":"aleks/varnish","Action":"pull"}]}]}
```
The registry config is:
```ruby
version: 0.1
log:
  fields:
    service: registry
storage:
    cache:
        layerinfo: inmemory
    filesystem:
        rootdirectory: /data/registry
http:
    addr: 0.0.0.0:5000
    debug:
       addr: 0.0.0.0:5001
auth:
  token:
    realm: https://portus.testcompany.co.uk/v2/token
    service: distribution.testcompany.co.uk
    issuer: portus.test.lan
    rootcertbundle: /registry/portus.crt
notifications:
  endpoints:
    - name: portus
      url: http://portus:3000/v2/webhooks/events
      timeout: 500ms
      threshold: 5
      backoff: 1s
```
It must be a misconfiguration. How am I supposed to setup communication between Portus and Registry so getting the manifest doesn't drop the 401 Unauthorised error?


Here's the tcpdump logs done on registry that confirms communication from Portus to registry that's passing through multiple reverse proxies is fine.

```bash
14:52:55.597860 IP 172.17.0.2.53759 > e9aad501d662.5000: Flags [P.], seq 1024993160:1024993597, ack 1962696117, win 115, options [nop,nop,TS val 515541309 ecr 515541309], length 437
E.....@.@.1........D....=.'.t.]....sZD.....
...=...=GET /v2/aleks/varnish/manifests/sha256:46b161cb075d45a947b458475b58afad5c99b3372b1e4b6976dc46cf7ab992e8 HTTP/1.1
Host: distribution.testcompany.co.uk
X-Real-IP: 10.80.1.118
X-Forwarded-Proto: https
X-Source: nginx
Accept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3
Accept: */*
User-Agent: Ruby
X-Forwarded-For: 127.0.0.1
x-director: distribution_be
X-Backend-Host: distribution.testcompany.co.uk
X-Varnish: 289191431


14:52:55.604535 IP e9aad501d662.5000 > 172.17.0.2.53759: Flags [P.], seq 1:509, ack 437, win 122, options [nop,nop,TS val 515541316 ecr 515541309], length 508
E..0..@.@......D........t.].=.)=...zZ......
...D...=HTTP/1.1 401 Unauthorized
Content-Type: application/json; charset=utf-8
Docker-Distribution-Api-Version: registry/2.0
Www-Authenticate: Bearer realm="https://portus.testcompany.co.uk/v2/token",service="distribution.testcompany.co.uk",scope="repository:aleks/varnish:pull"
Date: Mon, 28 Sep 2015 14:52:55 GMT
Content-Length: 172

{"errors":[{"code":"UNAUTHORIZED","message":"access to the requested resource is not authorized","detail":[{"Type":"repository","Name":"aleks/varnish","Action":"pull"}]}]}
```

**Comment by mssola** [comment]
Hi @morodin, Portus inserting a "1" into the DB was a regression that has been fixed here: 71c84b87f8a77f1359c1641e213fcffd7c6ab45c. Moreover, this commit also improves the logging of Portus trying to fetch the manifest from the registry. Maybe this can help you debug why Portus is frozen ?

**Comment by mssola** [comment]
Hi @aleksnikolic. If I recall correctly you had upgraded from version 1.0.1. There were a ton of changes since that version. One thing we added was the "portus" special user, that is now the responsible for handling the communication between Portus and the registry. Make sure that this user exists on your DB and that has the proper credentials (take a look at how the [current secrets.yml](https://github.com/SUSE/Portus/blob/master/config/secrets.yml) looks like). If this is the case, you can create the "portus" user with the rake task portus:create.

Otherwise, this might be a misconfiguration. Make sure that you're using the proper ports for everything. As an exampe, you can find how we configure everything in the rpm [here](https://github.com/SUSE/Portus/tree/master/packaging/suse/portusctl/templates).

Moreover, I'm also working into providing an NGinx config and describe a working setup with it. It shouldn't take me too long but it's low priority :P

**Comment by morodin** [comment]
@mssola: The freezing problem stays. Shall i open another issue for this? It is really important, in the moment we can't run the registry/Portus environment at all. Perhaps you can give us an insight on why the problem occurs:

Whenever i try to push an image, Portus freezes exactly at the function get_response_token, line 140 `Net::HTTP.start`. The start call never returns. The trace till this point (with added debugging messages) is:

```
Registry client called, host=dregistry.example.com:5000, use_ssl=true, username=
Registry client manifest, repoistory=userx/busybox6, tag=sha256:4c537efbcc9382e8d6486b5185279f4c902532d4af3be707cff1a1cfb1bc01d8
perform_request called, uri=https://dregistry.example.com:5000/v2/userx/busybox6/manifests/sha256:4c537efbcc9382e8d6486b5185279f4c902532d4af3be707cff1a1cfb1bc01d8
perform_request: before calling get_response_token https://dregistry.example.com:5000/v2/userx/busybox6/manifests/sha256:4c537efbcc9382e8d6486b5185279f4c902532d4af3be707cff1a1cfb1bc01d8
get_response_token: uri=dregistry.example.com,5000, use_ssl=true
get_response_token: inside http.start
get_response_token: req done: #<Net::HTTP::Get:0x007fca97f9dcd8>
perform_request: after calling get_response_token res.code= 401
perform_request: 401
perform_request: calling request_auth_token
request_auth_token called, bearer_real=https://dregistry.example.com:3000/v2/token, query={"service"=>"dregistry.example.com:5000", "account"=>"portus", "scope"=>"repository:userx/busybox6:pull"}, uri=https://dregistry.example.com:3000/v2/token?account=portus&scope=repository%3Auserx%2Fbusybox6%3Apull&service=dregistry.example.com%3A5000
request_auth_token called, credentials found
request_auth_token: before calling get_response_token
get_response_token: uri=dregistry.example.com,3000, use_ssl=true
```

If i change the function a little bit and add `http.open_timeout`, it returns with the following error:

```
Could not fetch the tag for target {"mediaType"=>"application/vnd.docker.distribution.manifest.v1+json", "size"=>6709, "digest"=>"sha256:4c537efbcc9382e8d6486b5185279f4c902532d4af3be707cff1a1cfb1bc01d8", "length"=>6709, "repository"=>"userx/busybox6", "url"=>"https://dregistry.example:5000/v2/userx/busybox6/manifests/sha256:4c537efbcc9382e8d6486b5185279f4c902532d4af3be707cff1a1cfb1bc01d8"}
Reason: execution expired
```

Even more strange, when i call the manifest function from the rails console:
```
myClient=Portus::RegistryClient.new("dregistry.example:5000",true)
myClient.manifest("userx/busybox6","sha256:c3102a0622d47e661b61208c19de53826fc725f9840b915ad1afd17ee5be6f70")
```
... everything works as excepted. Do you have any hint?

**Comment by mssola** [comment]
Hi @morodin, I've tried to reproduce your issue by creating an NGinx environment, and I could finally reproduce this hang. However, I had some different bits than in your case (for example, the mediaType that I got was not the one from the manifest).

Therefore, I think it'd be better to open a new issue with all the info we can gather, since this is totally unrelated to the original issue :) Moreover, are you using NGinx as a proxy? If so, could you provide your configuration? (in the new issue)

Thanks for your patience! :)

**Comment by morodin** [comment]
I created a new issue: https://github.com/SUSE/Portus/issues/373 for the freezing problem.

**Comment by mssola** [comment]
@aleksnikolic please open up a new issue if you're still facing your issue.

I will close this issue since the discussion has diverged too much from the original issue, and the original issue seems to be fixed. If I'm wrong, feel free to re-open this issue. Thanks to everyone that has provided feedback about Portus' issues :)

**Comment by morodin** [comment]
@mssola The second issue is still open: `Showing /portus/app/views/repositories/show.html.slim where line #35 raised:`
```
undefined method `username' for nil:NilClass
td= tag.author.username
```
This happens, when there is a discrepancy between registry and Portus and the crono synchronization job is run.

**Comment by mssola** [comment]
@morodin oops, you're right :) I've open up a new issue for it #374. Expect a fix as soon as possible ;)
